### PR TITLE
🌈 Add multi-asset progress bars

### PR DESCRIPTION
**Pull request by czunker**
Signed-off-by: Christian Zunker <christian@mondoo.com>

### CODE REVIEW

**Comment by czunker** [comment]
For some examples, how the new progress bars look like: https://github.com/mondoohq/cnspec/pull/282#issuecomment-1380103645

**Review by vjeffrey — COMMENTED** [review]
tried this out locally, it looks like if we can't connect to all the assets the progress bars don't get filled in

Screenshot 2023-01-13 at 08 58 21

**Comment by czunker** [comment]
> tried this out locally, it looks like if we can't connect to all the assets the progress bars don't get filled in
> 
> Screenshot 2023-01-13 at 08 58 21

Yes, the progress bars show the progress of the scan of the assets. Progress is measured as finished queries over all the queries which apply. When we cannot scan the assets, there is no progress. This is the same as before, but now it has become more obvious.

File: apps/cnquery/cmd/scan.go [hunk 1]
```diff
@@ -267,6 +267,7 @@ This example connects to Microsoft 365 using the PKCS #12 formatted certificate:
 		cmd.Flags().MarkHidden("category")
 
 		// output rendering
+		cmd.Flags().Int("num-assets", 30, "Set number of assets for which progress is displayed. (0 will show all assets)")
```
Comment (chris-rock) [inline apps/cnquery/cmd/scan.go:270]: I think this is not something we want user to configure

**Comment by chris-rock** [comment]
Screenshot 2023-01-17 at 10 28 44

I agree with @vjeffrey

It makes sense that the individual asset bar stays gray but it needs to be indicated that an error happened. Therefore instead of

```

███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-controller-manager-minikube
███████████████████████████████████████████████████████████████████████████ 100% kube-system/coredns-64897985d-b2p8g
███████████████████████████████████████████████████████████████████████████ 100% kube-system/storage-provisioner
███████████████████████████████████████████████████████████████████████████ 100% kube-system/etcd-minikube
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░   0% kube-system/kube-proxy
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-scheduler-minikube
███████████████████████████████████████████████████████████████████████████ 100% K8s Cluster minikube
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-apiserver-minikube
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░   0% kube-system/coredns
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-proxy-fhk9n
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░   0% kube-system/coredns-64897985d


███████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░  73% overall 8/11 assets
```

It should look more like this:

```

███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-controller-manager-minikube
███████████████████████████████████████████████████████████████████████████ 100% kube-system/coredns-64897985d-b2p8g
███████████████████████████████████████████████████████████████████████████ 100% kube-system/storage-provisioner
███████████████████████████████████████████████████████████████████████████ 100% kube-system/etcd-minikube
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░    X kube-system/kube-proxy
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-scheduler-minikube
███████████████████████████████████████████████████████████████████████████ 100% K8s Cluster minikube
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-apiserver-minikube
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░    X kube-system/coredns
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-proxy-fhk9n
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░    X kube-system/coredns-64897985d


███████████████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░  100% 8/11 completed 3/11 errors
```

**Comment by chris-rock** [comment]
Not directly related to the progress bar, but with the error reporting in the summary we should not log those as warning anymore:
  
Screenshot 2023-01-17 at 10 33 53

The summary also should print the target and then the error underneath. Currently it is a bit strange that errors are rendered differently then results

Screenshot 2023-01-17 at 10 34 20

**Comment by czunker** [comment]
@vjeffrey / @chris-rock Please have another look.

The output now looks like this:
```
→ resolved assets resolved-assets=16

░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░    X K8s Cluster kind-kind
███████████████████████████████████████████████████████████████████████████ 100% local-path-storage/local-path-provisioner-684f458cdd
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kindnet-qfkfv
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kindnet
███████████████████████████████████████████████████████████████████████████ 100% kube-system/coredns-565d847f94
███████████████████████████████████████████████████████████████████████████ 100% local-path-storage/local-path-provisioner-684f458cdd-x7cwd
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-proxy-ht98w
███████████████████████████████████████████████████████████████████████████ 100% kube-system/coredns
███████████████████████████████████████████████████████████████████████████ 100% local-path-storage/local-path-provisioner
███████████████████████████████████████████████████████████████████████████ 100% kube-system/coredns-565d847f94-np68t
███████████████████████████████████████████████████████████████████████████ 100% kube-system/etcd-kind-control-plane
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-controller-manager-kind-control-plane
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-scheduler-kind-control-plane
███████████████████████████████████████████████████████████████████████████ 100% kube-system/coredns-565d847f94-4kktc
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-apiserver-kind-control-plane
███████████████████████████████████████████████████████████████████████████ 100% kube-system/kube-proxy


██████████████████████████████████████████████████████████████████████░░░░░  94% overall 15/16 assets 1/16 errors

Summary (16 assets)
========================
Target:     kube-system/coredns
Datapoints: 1

Target:     local-path-storage/local-path-provisioner
Datapoints: 1

Target:     kube-system/coredns-565d847f94-4kktc
Datapoints: 1

Target:     kube-system/kube-apiserver-kind-control-plane
Datapoints: 1

Target:     kube-system/kube-proxy-ht98w
Datapoints: 1

Target:     local-path-storage/local-path-provisioner-684f458cdd-x7cwd
Datapoints: 1

Target:     kube-system/kindnet
Datapoints: 1

Target:     kube-system/kindnet-qfkfv
Datapoints: 1

Target:     kube-system/kube-scheduler-kind-control-plane
Datapoints: 1

Target:     kube-system/coredns-565d847f94
Datapoints: 1

Target:     local-path-storage/local-path-provisioner-684f458cdd
Datapoints: 1

Target:     kube-system/etcd-kind-control-plane
Datapoints: 1

Target:     kube-system/kube-controller-manager-kind-control-plane
Datapoints: 1

Target:     kube-system/kube-proxy
Datapoints: 1

Target:     K8s Cluster kind-kind
✕ Error for asset K8s Cluster kind-kind: rpc error: code = InvalidArgument desc = asset does not match any of the activated query packs

Target:     kube-system/coredns-565d847f94-np68t
Datapoints: 1
```

**Comment by chris-rock** [comment]
Looks much better to me. Thank you @czunker

@tas50 Does this meet your expectations?

**Comment by czunker** [comment]
> @czunker is there a reason the scan doesn't happen from the first asset down? It's totally random which item scans first which makes for a weird scan visualization.
> 
> image

No, no specific reason. I changed the PR. so they are now ordered as scanned.

**Review by vjeffrey — APPROVED** [review]
looks great!

File: apps/cnquery/cmd/scan.go [hunk 2]
```diff
@@ -360,12 +363,19 @@ func getCobraScanConfig(cmd *cobra.Command, args []string, provider providers.Pr
...
-		IsIncognito:    viper.GetBool("incognito"),
-		DoRecord:       viper.GetBool("record"),
-		QueryPackPaths: viper.GetStringSlice("querypack-bundle"),
-		QueryPackNames: viper.GetStringSlice("querypacks"),
+		Features:          opts.GetFeatures(),
+		IsIncognito:       viper.GetBool("incognito"),
+		DoRecord:          viper.GetBool("record"),
+		QueryPackPaths:    viper.GetStringSlice("querypack-bundle"),
+		QueryPackNames:    viper.GetStringSlice("querypacks"),
+		ProgressNumAssets: progressNumAssets,
```
Comment (chris-rock) [inline apps/cnquery/cmd/scan.go:378]: ProgressNumAssets is a cli stearing property that has nothing to do with the scan job. In addition it should only configured once and is not for the job to decide

File: explorer/scan/cnquery_explorer_scan.proto [hunk 3]
```diff
@@ -12,5 +12,6 @@ message Job {
   
   bool do_record = 20;
   repeated string query_pack_filters = 21;
+  int32 progress_num_assets = 22;
```
Comment (chris-rock) [inline explorer/scan/cnquery_explorer_scan.proto:15]: We need to revert the proto change. This does not belong here

File: explorer/scan/local_scanner.go [hunk 4]
```diff
@@ -190,35 +192,70 @@ func (s *LocalScanner) distributeJob(job *Job, ctx context.Context, upstreamConf
...
+		progressProg = progress.NoopProgram{}
 	}
 
-	return reporter.Reports(), true, nil
+	scanGroup := sync.WaitGroup{}
+	scanGroup.Add(1)
+	finished := false
+	go func() {
+		defer scanGroup.Done()
+		defer progressProg.Quit()
```
Comment (chris-rock) [inline explorer/scan/local_scanner.go:216]: I do not understand why we have two defers that call `defer progressProg.Quit()`

**Review by chris-rock — COMMENTED** [review]
I apologize for the late feedback. Thank you Christian for the improvement and it is the right move into the right direction. Unfortunately we are not there yet.

- the amount of progress bars is not to be configured by the user
- the proto definition of the job should not have gotten a config for the amount of progress bars

In addition the progress bar does not work as specified yet:

- the overall progress does not lead to completion if individual assets fail
- the complete progress is not rounded properly. When I have 11 assets and 3 fail I see 73% completeness
- the bars do not stack up. In the spec we have 3 areas: completed progress, running progress and overall progress. At this point with just 1 parallel runner the output should be

File: explorer/scan/local_scanner.go (hunk 4 above)
Comment (czunker) [reply explorer/scan/local_scanner.go:216]: After our discussion, I reworked how the tea Program is closed, and this is no longer needed.

Implemented in https://github.com/mondoohq/cnquery/pull/823

**Pull request merged**